kclvm/sema/src/resolver/doc.rs

```diff
@@ -0,0 +1,583 @@
+        return self.data.iter().all(|x| x.trim().len() == 0);
+    }
+}
+
+/// remove the leading and trailing empty lines
+fn strip(doc: Vec<String>) -> Vec<String> {
+    let mut i = 0;
+    let mut j = 0;
+    for (line_num, line) in doc.iter().enumerate() {
+        if !line.trim().is_empty() {
+            i = line_num;
+            break;
+        }
+    }
+
+    for (line_num, line) in doc.iter().enumerate().rev() {
+        if !line.trim().is_empty() {
+            j = line_num;
+            break;
+        }
+    }
+
+    doc[i..j + 1].to_vec()
+}
+
+/// Checks if current line is at the beginning of a section
+fn is_at_section(doc: &mut Reader) -> bool {
+    doc.seek_next_non_empty_line();
+    if doc.eof() {
+        return false;
+    }
+    let l1 = doc.peek(0, true);
+    let l1 = l1.trim();
+    let l2 = doc.peek(1, true);
+    let l2 = l2.trim(); // ---------- or ==========
+    let l2_char_set = l2.chars().collect::<HashSet<char>>();
+
+    if l2.len() >= 3
+        && l2_char_set.len() == 1
+        && (l2.contains('-') || l2.contains('='))
+        && l1.len() != l1.len()
+    {
+        // todo: when line2 is conformed with "-" or "=", but the number of the "-/=" mismatch the section title length, mark as a section and return a warning
+        return false;
+    }
+    l2.starts_with(&"-".repeat(l1.len())) || l2.starts_with(&"=".repeat(l1.len()))
+}
+
+/// read lines before next section beginning, continuous empty lines will be merged to one
+fn read_to_next_section(doc: &mut Reader) -> Vec<String> {
+    let mut section = doc.read_to_next_empty_line();
+
+    while !is_at_section(doc) && !doc.eof() {
+        if doc.peek(1, false).trim().is_empty() {
+            section.push(doc.peek(1, false));
+        }
+        section.append(&mut doc.read_to_next_empty_line());
+    }
+    section
+}
+
+/// parse the Attribute Section of the docstring to list of Attribute
+fn parse_attr_list(content: String) -> Vec<Attribute> {
+    let mut r = Reader::new(content);
+    let mut attrs = vec![];
+    while !r.eof() {
+        let header = r.read();
+        let header = header.trim();
+        if header.contains(" : ") {
+            let parts: Vec<&str> = header.split(" : ").collect();
+            let arg_name = parts[0];
+
+            let desc_lines = r
+                .read_to_next_unindented_line()
+                .iter()
+                .map(|s| s.trim().to_string())
+                .collect();
+            attrs.push(Attribute::new(arg_name.to_string(), desc_lines));
+        } else {
+            r.read_to_next_unindented_line();
+        }
+    }
+    attrs
+}
+
+/// parse the summary of the schema. The final summary content will be a concat of lines in the original summary with whitespace.
+fn parse_summary(doc: &mut Reader) -> String {
+    if is_at_section(doc) {
+        // no summary provided
+        return "".to_string();
+    }
+    let lines = read_to_next_section(doc);
+    return lines
```

Comment:
Remove the return keyword. U can run `cargo clippy --fix --allow-dirty` to fix it automatically.